### ISSUE TITLE
DeepExpDoubleDuelQPlayer -> Unstable Learning

### ISSUE THREAD

**Issue by MostlyInsane** [fcarsten/tic-tac-toe#4]
Hey, 

First of all I would like to thank you on a very detailed tutorial on deep reinforcement learning.

Okay so what happened was I followed your guidelines and went ahead and coded a network very much similar to yours in PyTorch, so when I was training my network, during training it did good in between, but all of a sudden it seems to forget what it has learnt and leading to 100% losses in between and again learning and again leading to 100% losses and the cycle continues ? 

![Figure_1](https://user-images.githubusercontent.com/53304726/90371420-8f279280-e08c-11ea-9779-095a3021ba1b.png)

(Trained Against A MinMax Agent)

My configs:
->Network architecture same as yours
->I didn't use a one hot encode
->Updated my target network every 5 games (I tried changing it 4, 5 ,6, didn't make a difference)
->lr = 0.001, PreTrainingGames = 3000(changed to 500, 1000), Epsilon = 1, Discount =0.99, BatchSize =66

have you encountered such a pattern during your training process ?, if so what precautions have you taken to avoid such 'catastrophic forgetting' ?

Also I coudnt understand what 'tau' was in updating your target network, aren't you supposed to update based on number of episdoes(/Games), but in yours it was a float factor), could you provide some insight into this, Thanks a lot

regards,
Nikhil

**Comment by fcarsten** [comment]
I had similar things occur regularly while I developed the code for this tutorial. The final version however seems quite stable. There is still a non-zero change of it happening I think but I have not experienced it with my current network and code.

When it happened, it was very very hard to debug as there so many things that can go wrong. Some lessons I learned:
* Don't mess with the Maths. While the whole network looks like you can just plug&play like a lego set, this is not true.  The middle layers are usually fine to mess with, but don't change the input and output layers. The ones that are save to change are the "hidden" layers in my diagram. The other ones need to be the way they are or the underlying theory no longer works.
* Hyperparameter make a difference. Especially if some values are initialized much too large or small.
* Sometimes it's the python code: mis-alingned arrays, incorrect expressions, involuntary truncation to integer, ...

What are you using instead of one hot encode?

Re tau: When updating the target network, I don't just replace it with the main network. The 2 networks are combined and tau determines the respective weight for combining them:
   ` op_holder.append(t.assign((s.value() * tau) + ((1 - tau) * t.value())))`

**Comment by MostlyInsane** [comment]
Hey,

Thanks a lot for taking time from your daily schedule and helping me out,

- My input and the output layers are the same like yours so I don't think that that's the issue,
- I have spent quite a bit of time tuning the hyper-parameters, but the model didn't seem to get any 
  better
- I am thinking its probably some bug in code that's causing this, ill re-code the entire thing while printing 
  out results at checkpoints and try to find out where I went wrong.

As for the one hot encoder, I felt, you just need to remember the move you played right ?! (for updating the **Main-Networks** weights), so I just logged the 'Indices of the Action played for that particular state' (i.e 0-8, where 0 is top left, 8 is bottom right) and then during training I cloned the predicted 'QValues' from the **Main-Network** for that state and modified the QValue (In the **cloned** prediction) at the ActionIndex (i.e 0-8) and calculated the Loss->Backpropagate->UpdateWeights followed by updating my target network after every 5 games.

Thanks for the explanation on tau, is there any particular reason you are updating the target network like that?! I would really appreciate some more intuition on it (I hope I am not making your daily life harder :P, only if you are free), I recently got into RL, I usually work with Image processing and other SML stuff so I don't really know much in this field.

Regards,
Nikhil
